note command

DaeguDude opened the pull request:
Adds the command `/note` or `/notes` that prints the link to the thoughts on note taking.
These thoughts are collected from TOP discord server I believe, and someone has made it as a google document.

https://docs.google.com/document/d/1fas8xwMRi4GVrnCtp9U5OL7Y9QB6v5H_Wx19m4UQwgI/edit

I have seen a lot of people asking about note taking, so I thought this could be useful.

twalton83 commented:
I appreciate you adding this command! I have a small ask, we changed the formatting of the commands slightly. An example of this can be found [here](https://github.com/TheOdinProject/odin-bot-v2/blob/odin-bot-tests/botCommands/callbacks.js). Additionally, we will be making it mandatory to add tests to any new commands to be added. An example of the tests can be seen [here](https://github.com/TheOdinProject/odin-bot-v2/blob/odin-bot-tests/botCommands/callbacks.test.js). It would be super helpful if you could perhaps consider at minimum refactoring the command to the new format, and super super helpful if you could add the spec file as well! Ultimately you can copy and paste the tests and just replace the callbacks strings with the appropriate note/notes strings. Let me know if you have any questions.

DaeguDude commented:
> It would be super helpful if you could perhaps consider at minimum refactoring the command to the new format

Yes of course I can do it! But I am a bit confused if you mean just my command or all you want me to work on other commands as well?

> and super super helpful if you could add the spec file as well! Ultimately you can copy and paste the tests and just replace the callbacks strings with the appropriate note/notes strings

I actually haven't gone through the test section yet, so I don't know how to implement it yet. But if this isn't urgent thing,I can try to go through basic materials of test and implement it? Or should I just copy and paste the strings like you said? :)

twalton83 commented:
There’s going to be a totally new README with instructions, you won’t need any testing experience :) 

That said I’m totally happy to do it, though the bot test branch should be merged to master today I believe, and we really made it to where people don’t have to do much more than copying and pasting and changing the command name in the tests. It’s super important since as you might have seen, our commands are super out of wack as far as regex.

twalton83 commented:
The updated README has been merged, with test directions :) 

https://github.com/TheOdinProject/odin-bot-v2/blob/master/README.md

DaeguDude commented:
Okay! I'll take a look at make a change today. I'm not sure if I can do it in the right way, please let me know if I do something stupid.

twalton83 commented:
That’s what the pull request and review process is for ;)

DaeguDude commented:
So, I followed the instruction and everything went well but test. `npm test` creates a lot of fails, I am not too sure what I did wrong. Should I upload the error log?

Here's the brief look of it.
![Uploading Screen Shot 2021-01-11 at 6.13.23 PM.png…]()

twalton83 commented:
That's okay- it's the other tests and they should fail because our regexes are incorrect. If you run `npm test note.test.js` or however you named the test file, it'll run *those* tests only! If those pass, I'm happy to merge if you provide a screenshot.

DaeguDude commented:
Hi! I've just finished note command and its test file, if there's something to be fixed please let me know!

![Screen Shot 2021-01-13 at 11 46 10 AM](https://user-images.githubusercontent.com/35420073/104400069-4b7e7980-5595-11eb-870d-59bd8eab6cbf.png)

twalton83 commented:
@I3uckwheat Do you mind reviewing and merging this? Thank you!

I3uckwheat commented:
I don't think we want to be linking to someone else's google document. That has some inherent risk we do not want to take on.

I3uckwheat commented on botCommands/note.js:
```diff
@@ -0,0 +1,12 @@
+const { registerBotCommand } = require("../botEngine.js");
+
+const command = {
+  regex: /(?<![\w`!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?~])\/note(?![\w`!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?~)])/,
```
This regex is overly complicated. Take a look at a different example to get a better idea on how to do this. This is unreadable to a normal human and could take a LONG time to understand if there's a bug within.

I3uckwheat replied:
I also believe we should call it "notes" instead of "note". Nobody just makes one note.

I3uckwheat replied:
https://github.com/TheOdinProject/odin-bot-v2/blob/master/botCommands/callbacks.js

See here for example.

It's often very useful to look at the patterns established by other pieces of code before trying to establish your own.

DaeguDude replied:
> This regex is overly complicated. Take a look at a different example to get a better idea on how to do this. This is unreadable to a normal human and could take a LONG time to understand if there's a bug within.
Yes, I agree it is overly complicated :( I was having a hard time to pass all the tests that were provided and that one above was the one I came with that could pass all the tests. 

> I also believe we should call it "notes" instead of "note". Nobody just makes one note.
I will change it to the 'notes' instead of 'note'.

> It's often very useful to look at the patterns established by other pieces of code before trying to establish your own.
Yes, I initially went with this one actually, but it couldn't pass the tests if it's followed by any non word characters like '/', '.', '*', and etc. Any ideas how I can do it??? Or should I just leave it for the time being?

For example, `/\B\/notes` will not pass '/notes!', because it's followed by '!'.

I3uckwheat commented on botCommands/note.js:
```diff
@@ -0,0 +1,12 @@
+const command = {
+  regex: /(?<![\w`!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?~])\/note(?![\w`!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?~)])/,
+  cb: () => { 
+    return `To note, or not to note, that is the question. Please read this: https://docs.google.com/document/d/1fas8xwMRi4GVrnCtp9U5OL7Y9QB6v5H_Wx19m4UQwgI/edit`;
```
We really shouldn't be linking to a google doc, as it's very mutable. What happens when someone turns it into an advertisement and all our links from the bot in the past now link to that ad? Or worse.

I3uckwheat replied:
A better solution to this would be to make a markdown file, or link to the discord conversation that someone trustworthy retains ownership of.

DaeguDude replied:
Okay. For the time being, I guess I will link to the discord conversation by Briggs.

https://discord.com/channels/505093832157691914/505093832157691916/768161823366578176

DaeguDude replied:
Updated!

I3uckwheat commented on botCommands/note.js:
```diff
@@ -0,0 +1,12 @@
+
+registerBotCommand(command.regex, command.cb);
+
+module.exports = command;
```
Please add a newline at the end of this file

DaeguDude replied:
Got it

I3uckwheat commented on botCommands/note.test.js:
```diff
@@ -0,0 +1,52 @@
+      expect(command.cb()).toMatchSnapshot()
+    })
+  })
+})
```
Please add a newline at the end of this file

DaeguDude replied:
Okay.

DaeguDude replied:
Both of files are updated with a new line.

DaeguDude commented:
<img width="1214" alt="Screen Shot 2021-01-17 at 1 33 30 PM" src="https://user-images.githubusercontent.com/35420073/104831116-1477d300-58c9-11eb-8ee4-d40e4dfde2c9.png">

Currently it passes most of the tests, however when it's preceded or followed by any characters or special characters, it still matches the word. 

For example,

'Here is the thoughts on /notes!'
'Have a look at this a/notes'
'Have a look at this /notesb'

All these examples would still invoke the note command. Maybe this could be improved? Or are we okay with it?

twalton83 commented:
Tests must pass. For example, this is a huge issue with our /google command. If you drop https://google.com/searchsomething in the chat, the google command fires off. Same with /code, if you drop a URL or something with "/code" in it, the code command fires off.

I3uckwheat commented:
That should be preventing pushes, but we need to fix all the other tests first

DaeguDude commented:
I've updated my solution using negative lookahead and negative lookbehind.
Updated solution indicates that it shouldn't be followed any non-whitespace character(e.x 1, a, #, c, !).

Can you review them?

twalton83 commented:
Go ahead and post the passing tests as well :)

DaeguDude commented:
<img width="1232" alt="Screen Shot 2021-01-21 at 8 20 38 AM" src="https://user-images.githubusercontent.com/35420073/105252843-9e71b580-5bc1-11eb-99b5-39438835f967.png">
There you go :)

twalton83 commented:
Looks good to me, @I3uckwheat?

I3uckwheat commented:
Thanks for this! I really appreciate the work to fix everything and do the tests!
